protocol/custom-gas-token.md

```diff
@@ -0,0 +1,459 @@
+
+## Problem Statement + Context
+
+The prior CGT design anchored the gas token to an L1 ERC‑20, hard‑coding bridge paths, and cluttering core contracts with token metadata. These restrictions make it hard to handle the following scenarios properly:
+
+- Integrates with a token that doesn’t live on L1.
+- Launch chains whose native asset does not yet exist or have an ERC20 representation somewhere.
+- Experiment with custom supply management or economics.
+- Implement any novel or non-conventional bridging mechanisms.
+
+Additionally, this design is interested in feasibly becoming a standard version for future compatibility, such as interoperability. The new design needs to keep the OP Stack lean and become future-proof.
+
+## Proposed Solution
+
+### Design Principles
+
+The presented design has been thought out based on the following principles:
+
+- Stick with a minimal set of changes.
+- Remain token-agnostic and completely flexible for any use cases.
+- Preserve OP Stack [foundational properties](https://specs.optimism.io/background.html#foundations), which also make Stage 2 requirements achievable.
+
+### Architecture
+
+**`isCustomGasToken()` in L1 and L2**
+
+The `isCustomGasToken()` view function becomes the unique one that brings awareness to the system about whether or not the CGT mode is being used. It is a boolean flag that other functions in other contracts use to check if they can receive ETH values or not.
+
+In L1 placed in `OptimismPortal`, it will block `depositTransaction` calls that contain `msg.value`. The same will occur in their pair functions `sendMessage` in `L1CrossDomainMessenger` and ETH bridging in `L1StandardBridge`. The `SystemConfig` reads this setting from the `OptimismPortal`, since it is a relevant configuration parameter for the chain.
+
+In L2 placed in `L1Block`, it will block `initiateWithdrawal` call, which contains `msg.value` in `L2ToL1MessagePasser`. The same will happen in `sendMessage` in `L2CrossDomainMessenger` and ETH bridging methods in `L2StandardBridge` and `FeeVaults`.
+
+As a consequence, native asset mints and burns are decoupled from system transactions (deposits) in normal operations, and those are moved into the application layer with the new introduced predeploys.
+
+**`NativeAssetLiquidity` and `LiquidityController` Predeploys**
+
+A very large amount of native assets is pre-minted and stored in `NativeAssetLiquidity`, which is managed through the `LiquidityController`, which has the rights to withdraw and deposit to this contract.
+
+A code example of both contracts would look like this:
+
+```solidity
+contract NativeAssetLiquidity {
```

Comment:
If the only contract that can call this contract is the LiquidityController, does it make sense for the contracts to be separate?

Reply:
The general idea of separating them was to reduce the amount of logic in the liquidity contract to the most bare minimum given if there is a bug in it, its game over because an extreme amount of native asset could be minted

Reply:
I agree that the fewer lines of code, the lower the probability of error. However, having separate contracts means every mint and burn takes more gas. Also, given the LiquidityController completely controls the NativeAssetLiquidity contract, it will come down to the security of the LiquidityController contract.

Reply:
You are definitely right that it takes up more gas, its a marginal amount of gas, but this sort of conversation can only be resolved with a shared optimization target